### PR TITLE
updates for budget and finalize merkato

### PR DESCRIPTION
**Pull request by mark-antony1**
-error check for valid instantiated merkato
   -and if no errors insert new merkato in DB 

-return all dictionaries of the merkatos

### CODE REVIEW

**Comment by mark-antony1** [comment]
@nasaWelder

**Comment by nasaWelder** [comment]
I think there's one more thing we need to confirm:

can we start a merkato if we only own one of the coins?
 @livinginformation @15chrjef

File: db_init_script.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    coin = "ETH"
+    spread = ".1"
+    merkato_UUID = configuration['exchange'] + "coin={}_base={}".format(coin,base)
+    merkato = Merkato(configuration, coin, base, spread, merkato_UUID)
+    merkatos = get_all_merkatos()
+    complete_merkato_configs = generate_complete_merkato_configs(merkatos)
+    print(complete_merkato_configs)
+
+    print(merkatos)
+    merkato.exchange.get_all_orders()
```
Comment (nasaWelder) [inline db_init_script.py]: print the result

File: merkato/constants.py [hunk 2]
```diff
@@ -7,13 +7,13 @@
 known_exchanges = {
     'tux': 'tux',
     'polo': 'polo',
-    'bit': 'bit'
+    'bit': 'bit',
+    'test': 'test'
 }
 implemented_exchanges = ["tux"]
```
Comment (nasaWelder) [inline merkato/constants.py]: add test to implemented_exchanges

File: merkato/merkato.py [hunk 3]
```diff
@@ -2,24 +2,26 @@
...
+from merkato.utils.database_utils import update_merkato, insert_merkato
+from merkato.exchanges.tux_exchange.utils import translate_ticker
+from merkato.utils import create_price_data, validate_merkato_initialization, get_relevant_exchange
 from math import floor
 
 DEBUG = True
 
 
 class Merkato(object):
-    def __init__(self, configuration, exchange, coin, base, spread, ask_budget, bid_budget):
```
Comment (nasaWelder) [inline merkato/merkato.py:14]: budgets aren't needed to instantiate? that will change the GUI, which is fine, but how do you specify how much to spend on each merkato?

File: merkato/merkato.py [hunk 4]
```diff
@@ -2,24 +2,26 @@
...
-        self.exchange = TuxExchange(configuration, coin=coin, base=base)
-        self.mutex_UUID = configuration['exchange'] + "coin={}_base={}".format(coin,base)
+    def __init__(self, configuration, coin, base, spread, UUID):
```
Comment (nasaWelder) [inline merkato/merkato.py]: I dont think the user(GUI) will know the UUID format, I think it should still be calculated in init

Comment (nasaWelder) [reply merkato/merkato.py]: it's derived information

**Comment by needmoney90** [comment]
Behavior for a merkato with only one asset is currently undefined, I've been toying around with how to do it. I'm leaning towards just making the user MM from that point onwards, if the price goes below (or above) a region with orders, the bot will lie idle. Later I can figure out some scheme where profits from sales get moved lower down, but that will take some modeling and thinking about before I'm comfortable picking an option.

File: merkato/merkato.py (hunk 3 above)
Comment (needmoney90) [reply merkato/merkato.py:14]: After speaking yesterday, @15chrjef and I decided only reserve and the amount on the books are necessary for operation, 'budget' can be queried by updating a merkato and adding together the reserve + orders on the books.

Comment (needmoney90) [reply merkato/merkato.py:14]: Basically it was a duplicate field

File: merkato/constants.py (hunk 2 above)
Comment (needmoney90) [reply merkato/constants.py]: Ah, yes

File: merkato/merkato.py (hunk 3 above)
Comment (nasaWelder) [reply merkato/merkato.py:14]: so how do i start a new one and tell it to only use 30% of my idle BTC? @livinginformation

File: db_init_script.py (hunk 1 above)
Comment (mark-antony1) [reply db_init_script.py]: its left over from the previous bot file, not needed for any specific purpose

File: merkato/merkato.py (hunk 4 above)
Comment (mark-antony1) [reply merkato/merkato.py]: sure

**Comment by mark-antony1** [comment]
All set @nasaWelder?

**Comment by nasaWelder** [comment]
I guess I can test it. I'm out for a bit.

**Comment by nasaWelder** [comment]
add this script somewhere:


    from merkato.merkato import Merkato

    from merkato.utils.database_utils import *

    if no_merkatos_table_exists():
        create_merkatos_table()
    if not get_all_merkatos():
        Merkato(configuration={"privatekey": "s", "publickey": "qw", "limit_only":1, "exchange":"test"},coin = "XMR", base="BTC", spread=.07)


    for merk in get_all_merkatos():
        print(merk)
    
    print("all finished")

**Comment by nasaWelder** [comment]
~~1. the result of get_all_merkatos() needs to be list of dictionaries, it's now a list of tuples.~~
~~2. when i ran it, the auth keys weren't in there. and there was no config dictionary therein~~

remember, i want to be able to do 
    actives = [Merkato(**args) for args in get_all_merkatos()]

**Comment by nasaWelder** [comment]
I didn't see generate_complete_merkato_configs
extra arg needs to be removed: UUID

**Comment by nasaWelder** [comment]
File "./test_merk.py", line 16, in <module>
    for merk in generate_complete_merkato_configs(get_all_merkatos()):
  File "/home/manbearpig/bin/trash/merkato/merkato/utils/__init__.py", line 61, in generate_complete_merkato_configs
    exchange = get_exchange_from_db(tuple[0])
  File "/home/manbearpig/bin/trash/merkato/merkato/utils/database_utils.py", line 126, in get_exchange
    c.execute('''SELECT * FROM exchanges WHERE exchange = ?''', (exchange,))
sqlite3.OperationalError: no such table: exchanges

That error should probably be handled implicitly

**Comment by mark-antony1** [comment]
This is too disorganized for comments on a pull request, when can you get on discord?

File: merkato/merkato.py (hunk 3 above)
Comment (nasaWelder) [reply merkato/merkato.py:14]: @livinginformation my point is, I can't currently launch a merkato and watch it work to see if TestExchange is good to go.

**Comment by mark-antony1** [comment]
@nasaWelder Is this ready for approval? @livinginformation ready for merging?

**Comment by nasaWelder** [comment]
I approve since we can't test it without a working merkato class

**Pull request merged**